library/alloc/src/sync.rs

```diff
@@ -1300,6 +1352,12 @@ impl<T: ?Sized> Clone for Arc<T> {
     /// ```
     #[inline]
     fn clone(&self) -> Arc<T> {
+        // Check if we're locked. Relaxed is fine here because if this branch
+        // is taken, we're the only possible strong reference.
+        if self.inner().strong.load(Relaxed) == 0 {
+            panic!("Not allowed to clone a locked Arc");
+        }
```

Comment:
This doesn't need to be here since there's no way you can access an `Arc` while it is locked:
- the `Arc` you locked is mutably borrowed
- there was no other `Arc` because `lock_strong_count` required the strong count to be 1, meaning the `Arc` that locked it was the only one
- `Weak`s can't be upgrade in the meantime because the strong count being 0 will make them fail

@Aaron1011 for this same reason `Drop` is still sound, because it will never run while an `Arc` is locked

Reply:
We don't return a guard. We return a bool.

Reply:
Oh, I didn't see the methods were public. I was expecting them to be encapsulated in a public API like the one shown in the first post.

If this is the case then `Drop` is indeed leaky

Reply:
The guards aren't necessarily useful, since you still can't do anything with the guard (safely) you couldn't already do with the Arc (due to variance). However, these primitives would be useful to writing higher level guard types that, through a wrapper such as `Static` (or with a `T: 'static` bound) could provide safe, statically checked, mutable access to an `Arc`.

Reply:
> The guards aren't necessarily useful, since you still can't do anything with the guard (safely) you couldn't already do with the Arc (due to variance).

The fundamental idea behind this being able to provide an API akin to `Arc::get_mut` but that doesn't fail when `Weak`s still exists, instead temporarely prevents them from up upgrading. This is not possible with the current `Arc`.

Reply:
This check must be here no matter what, because, assuming a hypothetical `borrow_mut` that returns a guard, without this, you could cause UB in safe code by leaking the guard.